http-client/src/main/java/com/facebook/airlift/http/client/thrift/ThriftResponseHandler.java

```diff
@@ -0,0 +1,146 @@
+import com.facebook.drift.codec.ThriftCodec;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableListMultimap;
+import com.google.common.collect.ListMultimap;
+import com.google.common.net.MediaType;
+
+import javax.annotation.Nullable;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.UncheckedIOException;
+import java.util.List;
+
+import static com.facebook.airlift.http.client.thrift.ThriftProtocol.BINARY;
+import static com.google.common.net.HttpHeaders.CONTENT_TYPE;
+
+public class ThriftResponseHandler<T>
+        implements ResponseHandler<ThriftResponseHandler.ThriftResponse, RuntimeException>
+{
+    private final ThriftCodec<T> thriftCodec;
+
+    public ThriftResponseHandler(ThriftCodec<T> thriftCodec)
+    {
+        this.thriftCodec = thriftCodec;
+    }
+
+    @Override
+    public ThriftResponse handleException(Request request, Exception exception)
+    {
+        throw ResponseHandlerUtils.propagate(request, exception);
+    }
+
+    @Override
+    public ThriftResponse handle(Request request, Response response)
+    {
+        try {
+            return new ThriftResponse(response.getStatusCode(), response.getStatusMessage(), response.getHeaders(), thriftCodec, response.getInputStream());
+        }
+        catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    public class ThriftResponse
+    {
+        private final int statusCode;
+        private final String statusMessage;
+        private final ListMultimap<HeaderName, String> headers;
+        private final T value;
+        private final IllegalArgumentException exception;
+
+        ThriftResponse(int statusCode, String statusMessage, ListMultimap<HeaderName, String> headers, ThriftCodec<T> thriftCodec, InputStream thriftBytes)
+        {
+            this.statusCode = statusCode;
+            this.statusMessage = statusMessage;
+            this.headers = ImmutableListMultimap.copyOf(headers);
+            T value = null;
+            IllegalArgumentException exception = null;
+            ThriftProtocol thriftProtocol = getThriftProtocol(headers);
+            try {
+                value = ThriftProtocolUtils.read(thriftCodec, thriftProtocol, thriftBytes);
+            }
+            catch (Exception e) {
+                exception = new IllegalArgumentException("Unable to create " + thriftCodec.getType() + " from THRIFT response", e);
+            }
+            this.value = value;
+            this.exception = exception;
+        }
+
+        private ThriftProtocol getThriftProtocol(ListMultimap<HeaderName, String> headers)
+        {
+            ThriftProtocol thriftProtocol = BINARY;
+            HeaderName contentTypeHeader = HeaderName.of(CONTENT_TYPE);
+            if (headers.containsKey(contentTypeHeader)) {
+                List<String> headerParameters = headers.get(contentTypeHeader);
+                if (headerParameters.size() > 0) {
```

Comment:
Let's check that there's only a singel `Content-Type` header. Request containing more then a single one is probably invalid